JsonSerializer cannot serialize "long"

On MacOS with AppleClang 13:
```
In file included from /Users/david/Documents/HZI/memilio/cpp/examples/read_graph.cpp:20:
In file included from /Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/mobility/mobility.h:30:
In file included from /Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/epidemiology/contact_matrix.h:24:
In file included from /Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/epidemiology/damping.h:24:
In file included from /Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/utils/type_safe.h:23:
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:778:5: error: no matching function for call to 'serialize_internal'
    serialize_internal(io, t);
    ^~~~~~~~~~~~~~~~~~
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/json_serializer.h:621:14: note: in instantiation of function template specialization 'mio::serialize<mio::JsonContext, long>' requested here
        mio::serialize(ctxt, value);
             ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:593:9: note: in instantiation of function template specialization 'mio::JsonObject::add_element<long, nullptr>' requested here
    obj.add_element("Rows", mat.rows());
        ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:778:5: note: in instantiation of function template specialization 'mio::serialize_internal<mio::JsonContext, Eigen::Matrix<double, -1, -1, 0>>' requested here
    serialize_internal(io, t);
    ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/json_serializer.h:621:14: note: in instantiation of function template specialization 'mio::serialize<mio::JsonContext, Eigen::Matrix<double, -1, -1, 0>>' requested here
        mio::serialize(ctxt, value);
             ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/epidemiology/contact_matrix.h:215:13: note: in instantiation of function template specialization 'mio::JsonObject::add_element<Eigen::Matrix<double, -1, -1, 0>, nullptr>' requested here
        obj.add_element("Baseline", get_baseline());
            ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:674:7: note: (skipping 22 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
    t.serialize(io);
      ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:674:7: note: in instantiation of function template specialization 'mio::SecirModel::serialize<mio::JsonSerializer>' requested here
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:778:5: note: in instantiation of function template specialization 'mio::serialize_internal<mio::JsonSerializer, mio::SecirModel, nullptr>' requested here
    serialize_internal(io, t);
    ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/json_serializer.h:532:5: note: in instantiation of function template specialization 'mio::serialize<mio::JsonSerializer, mio::SecirModel>' requested here
    serialize(js, t);
    ^
/Users/david/Documents/HZI/memilio/cpp/models/secir/../secir/secir_parameters_io.h:73:42: note: in instantiation of function template specialization 'mio::serialize_json<mio::SecirModel>' requested here
        BOOST_OUTCOME_TRY(js_node_model, serialize_json(node.property, ioflags));
                                         ^
/Users/david/Documents/HZI/memilio/cpp/examples/read_graph.cpp:142:30: note: in instantiation of function template specialization 'mio::write_graph<mio::SecirModel>' requested here
    auto write_status = mio::write_graph(graph, "graph_parameters");
                             ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:637:6: note: candidate template ignored: requirement 'std::is_enum<long>::value' was not satisfied [with IOContext = mio::JsonContext, E = long]
void serialize_internal(IOContext& io, E e)
     ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:672:6: note: candidate template ignored: requirement 'is_expression_valid<serialize_t, mio::JsonContext, long>::value' was not satisfied [with IOContext = mio::JsonContext, T = long]
void serialize_internal(IOContext& io, const T& t)
     ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:727:6: note: candidate template ignored: requirement 'is_expression_valid<details::compare_iterators_t, long>::value' was not satisfied [with IOContext = mio::JsonContext, Container = long]
void serialize_internal(IOContext& io, const Container& container)
     ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:590:6: note: candidate template ignored: could not match 'EigenBase<type-parameter-0-1>' against 'const long'
void serialize_internal(IOContext& io, const Eigen::EigenBase<M>& mat)
     ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/io.h:560:6: note: candidate template ignored: requirement 'is_expression_valid<details::tuple_size_value_t, long>::value' was not satisfied [with IOContext = mio::JsonContext, Tup = long]
void serialize_internal(IOContext& io, const Tup& tup)
     ^
/Users/david/Documents/HZI/memilio/cpp/memilio/../memilio/io/json_serializer.h:445:17: note: candidate template ignored: requirement 'JsonType<long, void>::value' was not satisfied [with T = long]
    friend void serialize_internal(JsonContext& io, const T& t)
                ^
```

--- a/cpp/memilio/io/json_serializer.h
+++ b/cpp/memilio/io/json_serializer.h
@@ -104,32 +104,34 @@ struct JsonType<T, std::enable_if_t<conjunction_v<is_small_integral<T>, std::is_
         return Json::Value(Json::UInt(i));
     }
 };
+template <class T>
+using is_64bit_integral = std::integral_constant<bool, (std::is_integral<T>::value && sizeof(T) == 8)>;
 //signed big ints
-template <>
-struct JsonType<int64_t> : std::true_type {
-    static IOResult<int64_t> transform(const Json::Value& js)
+template <class T>
+struct JsonType<T, std::enable_if_t<conjunction_v<is_64bit_integral<T>, std::is_signed<T>>>> : std::true_type {
+    static IOResult<T> transform(const Json::Value& js)
     {
         if (js.isInt64()) {
             return success(js.asInt64());
         }
         return failure(StatusCode::InvalidType, "Json value is not a 64 bit integer.");
     }
-    static Json::Value transform(int64_t i)
+    static Json::Value transform(T i)
     {
         return Json::Value(Json::Int64(i));
     }
 };
 //unsigned big ints
-template <>
-struct JsonType<uint64_t> : std::true_type {
-    static IOResult<uint64_t> transform(const Json::Value& js)
+template <class T>
+struct JsonType<T, std::enable_if_t<conjunction_v<is_64bit_integral<T>, std::is_unsigned<T>>>> : std::true_type {
+    static IOResult<T> transform(const Json::Value& js)
     {
         if (js.isUInt64()) {
             return success(js.asUInt64());
         }
         return failure(StatusCode::InvalidType, "Json value is not an unsigned 64bit integer.");
     }
-    static Json::Value transform(uint64_t i)
+    static Json::Value transform(T i)
     {
         return Json::Value(Json::UInt64(i));
     }

--- a/cpp/tests/test_json_serializer.cpp
+++ b/cpp/tests/test_json_serializer.cpp
@@ -25,6 +25,10 @@
 #include "matchers.h"
 #include "distributions_helpers.h"
 #include "gtest/gtest.h"
+#include "json/config.h"
+#include <gtest/gtest-typed-test.h>
+#include <gtest/internal/gtest-type-util.h>
+#include <limits>
 #include <vector>
 
 namespace jsontest
@@ -112,31 +116,74 @@ struct DoubleList {
 
 } // namespace jsontest
 
-TEST(TestJsonSerializer, basic_type)
-{
+//checks round trip (de)serialization and intermediate value
+template<class T, class StoredType = T>
+void check_serialization_of_basic_type(T t)
+{        
     mio::JsonSerializer js;
-    serialize(js, 3);
-    ASSERT_EQ(js.value(), Json::Value(3));
+    serialize(js, t);
+    EXPECT_EQ(js.value(), Json::Value(StoredType(t)));
 
-    mio::JsonSerializer js2{Json::Value(3)};
-    auto r = deserialize(js2, mio::Tag<int>{});
+    mio::JsonSerializer js2{Json::Value(StoredType(t))};
+    auto r = deserialize(js2, mio::Tag<T>{});
     ASSERT_THAT(print_wrap(r), IsSuccess());
-    ASSERT_EQ(r.value(), 3);
+    EXPECT_EQ(r.value(), t);
 }
 
-TEST(TestJsonSerializer, basic_type_float)
+using SmallIntegerTypes = testing::Types<char, unsigned char, signed char, short, unsigned short, int, unsigned int>;
+template<class T>
+class TestJsonSerializerSmallInts : public testing::Test
+{};
+TYPED_TEST_SUITE(TestJsonSerializerSmallInts, SmallIntegerTypes);
+TYPED_TEST(TestJsonSerializerSmallInts, full_domain)
 {
+    //all small integers are stored as 32 bit int with the corresponding signedness
+    using StorageType = typename std::conditional<std::is_signed<TypeParam>::value, Json::Int, Json::UInt>::type;
+    check_serialization_of_basic_type<TypeParam, StorageType>(std::numeric_limits<TypeParam>::min());
+    check_serialization_of_basic_type<TypeParam, StorageType>(std::numeric_limits<TypeParam>::max());
+    check_serialization_of_basic_type<TypeParam, StorageType>(TypeParam(3));
+}
+
+using BigIntegerTypes = testing::Types<long, unsigned long, long long, unsigned long long>;template<class T>
+class TestJsonSerializerBigInts : public testing::Test
+{};
+TYPED_TEST_SUITE(TestJsonSerializerBigInts, BigIntegerTypes);
+TYPED_TEST(TestJsonSerializerBigInts, full_domain)
+{
+    using StorageType = typename std::conditional<std::is_signed<TypeParam>::value, Json::Int64, Json::UInt64>::type;
+    check_serialization_of_basic_type<TypeParam, StorageType>(std::numeric_limits<TypeParam>::min());
+    check_serialization_of_basic_type<TypeParam, StorageType>(std::numeric_limits<TypeParam>::max());
+    check_serialization_of_basic_type<TypeParam, StorageType>(TypeParam(3));
+}
+
+using FloatingPointTypes = testing::Types<float, double>;
+template<class T>
+class TestJsonSerializerFloats : public testing::Test
+{};
+TYPED_TEST_SUITE(TestJsonSerializerFloats, FloatingPointTypes);
+TYPED_TEST(TestJsonSerializerFloats, full_domain)
+{
+    //all floating points are stored as double
+    check_serialization_of_basic_type<TypeParam, double>(std::numeric_limits<TypeParam>::max());
+    check_serialization_of_basic_type<TypeParam, double>(std::numeric_limits<TypeParam>::min());
+    check_serialization_of_basic_type<TypeParam, double>(std::numeric_limits<TypeParam>::lowest());
+    check_serialization_of_basic_type<TypeParam, double>(TypeParam(3.12));
+}
+
+TEST(TestJsonSerializer, string)
+{
+    // TODO: currently does not work, single strings (i.e. not part of a composite class) 
+    // are wrongly stored as arrays of characters instead of json strings because
+    // the wrong (de)serialize_internal overload is selected by the compiler.
+    // check_serialization_of_basic_type(std::string());
+    // check_serialization_of_basic_type(std::string("Hello, World!"));
+
+    //round trip serialization still works without loss, but intermediate value is not as expected
     mio::JsonSerializer js;
-    serialize(js, 3.0f);
-    ASSERT_EQ(js.value(), Json::Value(3.0));
-
-    mio::JsonSerializer js2{Json::Value(3.0)};
-    auto r = deserialize(js2, mio::Tag<float>{});
+    mio::serialize(js, std::string("Hello"));
+    auto r = deserialize(js, mio::Tag<std::string>{});
     ASSERT_THAT(print_wrap(r), IsSuccess());
-    ASSERT_EQ(r.value(), 3.0f);
-
-    mio::JsonSerializer js3{Json::Value(1e200)};
-    ASSERT_THAT(print_wrap(deserialize(js3, mio::Tag<float>{})), testing::Not(IsSuccess()));
+    EXPECT_EQ(r.value(), "Hello");
 }
 
 namespace jsontest
@@ -151,14 +198,8 @@ enum class E : int
 
 TEST(TestJsonSerializer, enum)
 {
-    mio::JsonSerializer js;
-    serialize(js, jsontest::E::E1);
-    ASSERT_EQ(js.value(), Json::Value(1));
-
-    mio::JsonSerializer js2{Json::Value(1)};
-    auto r = deserialize(js2, mio::Tag<jsontest::E>{});
-    ASSERT_THAT(print_wrap(r), IsSuccess());
-    EXPECT_EQ(r.value(), jsontest::E::E1);
+    check_serialization_of_basic_type<jsontest::E, int>(jsontest::E::E0);
+    check_serialization_of_basic_type<jsontest::E, int>(jsontest::E::E2);
 }
 
 TEST(TestJsonSerializer, tuple)
@@ -220,14 +261,7 @@ DECL_TYPESAFE(int, TypeSafeInt);
 
 TEST(TestJsonSerializer, typesafe)
 {
-    mio::JsonSerializer js;
-    serialize(js, jsontest::TypeSafeInt(1));
-    ASSERT_EQ(js.value(), Json::Value(1));
-
-    mio::JsonSerializer js2{Json::Value(1)};
-    auto r = deserialize(js2, mio::Tag<jsontest::TypeSafeInt>{});
-    EXPECT_TRUE(r);
-    EXPECT_EQ(r.value(), jsontest::TypeSafeInt(1));
+    check_serialization_of_basic_type<jsontest::TypeSafeInt, int>(jsontest::TypeSafeInt(3));
 }
 
 namespace jsontest

